### PR TITLE
chore: add 1inch validator [SUP-7014]

### PR DESCRIPTION
**Pull request by sujithsomraaj**

### CODE REVIEW

File: src/forms/ERC4626FormImplementation.sol [hunk 1]
```diff
@@ -229,7 +228,7 @@ abstract contract ERC4626FormImplementation is BaseForm, LiquidityHandler {
         /// superform data
         if (
             vars.assetDifference * ENTIRE_SLIPPAGE
-            < singleVaultData_.amount * (ENTIRE_SLIPPAGE - singleVaultData_.maxSlippage)
+                < singleVaultData_.amount * (ENTIRE_SLIPPAGE - singleVaultData_.maxSlippage)
```
Comment (0xTimepunk) [inline src/forms/ERC4626FormImplementation.sol:231]: Please undo this change as it affects bytecode

File: src/forms/ERC4626FormImplementation.sol [hunk 2]
```diff
@@ -493,3 +491,6 @@ abstract contract ERC4626FormImplementation is BaseForm, LiquidityHandler {
         }
     }
 }
+
+/// 5268608643
```
Comment (0xTimepunk) [inline src/forms/ERC4626FormImplementation.sol:495]: Remove this

File: src/forms/ERC4626FormImplementation.sol (hunk 1 above)
Comment (0xTimepunk) [inline src/forms/ERC4626FormImplementation.sol:231]: Please undo this change as it affects bytecode

File: test/utils/BaseSetup.sol [hunk 3]
```diff
@@ -722,6 +726,11 @@ abstract contract BaseSetup is DSTest, StdInvariant, Test {
 
             contracts[vars.chainId][bytes32(bytes("DeBridgeForwarderMock"))] = vars.debridgeForwarderMock;
 
+            /// @dev 7.1.8 deploy OneInchMock. This mocks the beahvior of 1inch
```
Comment (0xTimepunk) [inline test/utils/BaseSetup.sol:729]: This is not consistent with the rest of the code. OneInchMock should be made persistent at the top of setup next to the other mocks. Here it should just grab its address into contracts mapping

File: script/utils/misc/run_script_mainnet_1inch.sh [hunk 4]
```diff
@@ -8,23 +8,31 @@ export POLYGON_RPC_URL=$(op read op://5ylebqljbh3x6zomdxi3qd7tsa/POLYGON_RPC_URL
...
+FOUNDRY_PROFILE=production forge script script/forge-scripts/misc/Mainnet.Deploy.OneInchValidator.s.sol:MainnetDeployOneInchValidator --sig "deployOneInchValidator(uint256,uint256,uint256)" 0 5 0 --rpc-url $OPTIMISM_RPC_URL --slow --account defaultKey --sender 0x48aB8AdF869Ba9902Ad483FB1Ca2eFDAb6eabe92 --legacy
+wait
+
+FOUNDRY_PROFILE=production forge script script/forge-scripts/misc/Mainnet.Deploy.OneInchValidator.s.sol:MainnetDeployOneInchValidator --sig "deployOneInchValidator(uint256,uint256,uint256)" 0 6 0 --rpc-url $BASE_RPC_URL --slow --account defaultKey --sender 0x48aB8AdF869Ba9902Ad483FB1Ca2eFDAb6eabe92 --legacy
+wait
+
+FOUNDRY_PROFILE=production forge script script/forge-scripts/misc/Mainnet.DeployOneInchValidator.s.sol:MainnetDeployOneInchValidator --sig "deployOneInchValidator(uint256,uint256,uint256)" 0 7 1 --rpc-url $FANTOM --slow --account defaultKey --sender 0x48aB8AdF869Ba9902Ad483FB1Ca2eFDAb6eabe92 --legacy
```
Comment (0xTimepunk) [inline script/utils/misc/run_script_mainnet_1inch.sh:37]: Think this file is missing the configure steps in prod

**Review by 0xTimepunk — COMMENTED** [review]
LGTM , few things to amend

File: src/crosschain-liquidity/1inch/OneInchValidator.sol [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+import { IBridgeValidator } from "src/interfaces/IBridgeValidator.sol";
+import { ISuperRegistry } from "src/interfaces/ISuperRegistry.sol";
+import { Error } from "src/libraries/Error.sol";
+import "src/vendor/1inch/IAggregationRouterV6.sol";
+
+/// @title OneInchValidator
+/// @dev Asserts OneInch txData is valid
+/// @author Zeropoint Labs
+
+/// @notice this does not import BridgeValidator as decodeSwapOutpuToken cannot have `pure` as visibility identifier
```
Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:13]: typo decodeSwapOutpuToken  -> decodeSwapOutputToken

File: src/vendor/1inch/IAggregationRouterV6.sol [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        IERC20 srcToken;
+        IERC20 dstToken;
+        address payable srcReceiver;
+        address payable dstReceiver;
+        uint256 amount;
+        uint256 minReturnAmount;
+        uint256 flags;
+    }
+
+    function swap(IAggregationExecutor executor, SwapDescription calldata desc, bytes calldata data) external payable;
```
Comment (gpersoon) [inline src/vendor/1inch/IAggregationRouterV6.sol:267]: could put the parameters below the function `swap()`  (or maybe this is formatted automatically)

File: src/crosschain-liquidity/1inch/OneInchValidator.sol [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        } else {
+            /// @dev 2. receiver address validation
+            /// @dev if withdraws, then receiver address must be the receiverAddress
+            if (receiver != args_.receiverAddress) revert Error.INVALID_TXDATA_RECEIVER();
+        }
+
+        /// @dev 3. token validations
+        if (args_.liqDataToken != fromToken) revert Error.INVALID_TXDATA_TOKEN();
+
+        return false;
```
Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:70]: doublecheck, is `hasDstSwap` indeed always `false`?
(the function `decodeDstSwap()` is implemented)

File: src/crosschain-liquidity/1inch/OneInchValidator.sol [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    /// returns useful parameters for validaiton
+    function _decodeTxData(bytes calldata txData_)
+        internal
+        view
+        returns (address fromToken, uint256 fromAmount, address toToken, address receiver)
+    {
+        bytes4 selector = bytes4(txData_[:4]);
+
+        /// @dev does not support any sequential pools with unoswap
+        /// NOTE: support UNISWAP_V2, UNISWAP_V3, CURVE, SHIBASWAP
```
Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:116]: `SHIBASWAP` isn't present in `enum Protocol {}` of `IAggregationRouterV6.sol`. Is this supported?

File: src/vendor/1inch/IAggregationRouterV6.sol [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+    function addressForPreTransfer(Address self) internal view returns (address) {
+        if (protocol(self) == Protocol.UniswapV2) {
+            return self.get();
+        }
+        return address(this);
+    }
+}
+
+/// @dev imported from https://docs.uniswap.org/contracts/v2/reference/smart-contracts/pair#token1
+interface IUniswapV2Pair {
```
Comment (gpersoon) [inline src/vendor/1inch/IAggregationRouterV6.sol:94]: This interface is used for both V2 and V3/ The both have the same functions so should be ok.
However might be good to rename the interface to a more generic name and perhaps add a comment.

File: src/crosschain-liquidity/1inch/OneInchValidator.sol [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+                if (toToken == fromToken) {
+                    toToken = IUniswapV2Pair(dex.get()).token1();
+                }
+            }
+            /// @dev if protocol is curve
+            else if (protocol == ProtocolLib.Protocol.Curve) {
+                uint256 toTokenIndex = (Address.unwrap(dex) >> _CURVE_TO_COINS_ARG_OFFSET) & _CURVE_TO_COINS_ARG_MASK;
+                toToken = ICurvePool(dex.get()).underlying_coins(int128(uint128(toTokenIndex)));
+            }
```
Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:139]: Although `enum procotol {}` only support 3 elements, it might be safer to add an else with an error

File: src/crosschain-liquidity/1inch/OneInchValidator.sol [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+            ProtocolLib.Protocol protocol = dex.protocol();
+
+            /// @dev if protocol is uniswap v2 or uniswap v3
+            if (protocol == ProtocolLib.Protocol.UniswapV2 || protocol == ProtocolLib.Protocol.UniswapV3) {
+                toToken = IUniswapV2Pair(dex.get()).token0();
+
+                if (toToken == fromToken) {
+                    toToken = IUniswapV2Pair(dex.get()).token1();
+                }
```
Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:133]: could check `fromToken()` is one of the pair tokens

File: src/crosschain-liquidity/1inch/OneInchValidator.sol [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+            fromToken = fromTokenUint256.get();
+            fromAmount = decodedFromAmount;
+            receiver = receiverUint256.get();
+
+            ProtocolLib.Protocol protocol = dex.protocol();
+
+            /// @dev if protocol is uniswap v2 or uniswap v3
+            if (protocol == ProtocolLib.Protocol.UniswapV2 || protocol == ProtocolLib.Protocol.UniswapV3) {
+                toToken = IUniswapV2Pair(dex.get()).token0();
```
Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:129]: note: malicious dex-es could return any value for `token0()` and `token1()`
similar for `underlying_coins()`

File: src/crosschain-liquidity/1inch/OneInchValidator.sol [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+                if (toToken == fromToken) {
+                    toToken = IUniswapV2Pair(dex.get()).token1();
+                }
+            }
+            /// @dev if protocol is curve
+            else if (protocol == ProtocolLib.Protocol.Curve) {
+                uint256 toTokenIndex = (Address.unwrap(dex) >> _CURVE_TO_COINS_ARG_OFFSET) & _CURVE_TO_COINS_ARG_MASK;
+                toToken = ICurvePool(dex.get()).underlying_coins(int128(uint128(toTokenIndex)));
```
Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:138]: doesn't check `dex.usePermit2()`

Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:138]: LifiValidator has special code for 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE
Maybe needed here too?

https://github.com/1inch/1inch-docs/blob/master/docs/educational-resources/beginner/faq-for-developers.mdx#faq-for-developers
Use the address 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE, 
this applies to the native asset of the chain, so MATIC on polygon, BNB on BNB chain, AVAX on Avalanche, etc.

File: src/crosschain-liquidity/1inch/OneInchValidator.sol [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        }
+        /// @dev decodes the generic router call
+        else if (selector == IAggregationRouterV6.swap.selector) {
+            (, IAggregationRouterV6.SwapDescription memory swapDescription, bytes memory extCallData) =
+                abi.decode(_parseCallData(txData_), (IAggregationExecutor, IAggregationRouterV6.SwapDescription, bytes));
+
+            fromToken = address(swapDescription.srcToken);
+            fromAmount = swapDescription.amount;
+            toToken = address(swapDescription.dstToken);
+            receiver = swapDescription.dstReceiver;
```
Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:149]: maybe also check:
- srcReceiver
- flags

File: src/crosschain-liquidity/1inch/OneInchValidator.sol (hunk 10 above)
Comment (gpersoon) [reply src/crosschain-liquidity/1inch/OneInchValidator.sol:152]: That is true; taking that a step further:
If you want you could even remove `if (protocol == ProtocolLib.Protocol.Curve)` to save some gas (and then add a comment to explain this)

File: src/crosschain-liquidity/1inch/OneInchValidator.sol (hunk 12 above)
Comment (gpersoon) [reply src/crosschain-liquidity/1inch/OneInchValidator.sol]: perhaps, depending on how easy it would be to add arbitrary dex addresses for users to use.
If users only use it from trusted sites then it would be ok;
but if a fake website advertises to use superform but adds a malicious dex this might not be ok

File: src/crosschain-liquidity/1inch/OneInchValidator.sol (hunk 13 above)
Comment (gpersoon) [reply src/crosschain-liquidity/1inch/OneInchValidator.sol:151]: could make sure this is false

File: script/forge-scripts/Abstract.Deploy.Single.s.sol [hunk 15]
```diff
@@ -140,7 +142,7 @@ abstract contract AbstractDeploySingle is BatchScript {
         "RewardsDistributor",
         "DeBridgeValidator",
         "DeBridgeForwarderValidator",
-        "AxelarImplementation"
+        "OneInchValidator" "AxelarImplementation"
```
Comment (0xTimepunk) [inline script/forge-scripts/Abstract.Deploy.Single.s.sol:145]: Isn't this a bug? There is no comma so I doubt this will run/compile

File: script/forge-scripts/Abstract.Deploy.Single.s.sol [hunk 16]
```diff
@@ -697,25 +704,30 @@ abstract contract AbstractDeploySingle is BatchScript {
...
         if (vars.chainId == BASE) {
-            uint8[] memory bridgeIdsBase = new uint8[](3);
+            uint8[] memory bridgeIdsBase = new uint8[](4);
             /// @dev this is the new id of lifi validator
             bridgeIdsBase[0] = 101;
 
             /// @dev these are debridge
             bridgeIdsBase[1] = 4;
             bridgeIdsBase[2] = 5;
+            bridgeIdsBase[3] = 6;
```
Comment (0xTimepunk) [inline script/forge-scripts/Abstract.Deploy.Single.s.sol:714]: could update comment above to indicate 6 is 1inch

File: script/utils/misc/run_script_mainnet_1inch.sh [hunk 17]
```diff
@@ -8,23 +8,57 @@ export POLYGON_RPC_URL=$(op read op://5ylebqljbh3x6zomdxi3qd7tsa/POLYGON_RPC_URL
...
 
-FOUNDRY_PROFILE=production forge script script/forge-scripts/misc/Mainnet.Deploy1inch.s.sol:MainnetDeploy1inch --sig "deploy1inch(uint256,uint256)" 0 0 --broadcast --rpc-url $ETHEREUM_RPC_URL --slow --account defaultKey --sender 0x48aB8AdF869Ba9902Ad483FB1Ca2eFDAb6eabe92
+FOUNDRY_PROFILE=production forge script script/forge-scripts/misc/Mainnet.Deploy.OneInchValidator.s.sol:MainnetDeployOneInchValidator --sig "deployOneInchValidator(uint256,uint256,uint256)" 0 0 0 --rpc-url $ETHEREUM_RPC_URL --slow --account defaultKey --sender 0x48aB8AdF869Ba9902Ad483FB1Ca2eFDAb6eabe92 --legacy
 wait
 
-FOUNDRY_PROFILE=production forge script script/forge-scripts/misc/Mainnet.Deploy1inch.s.sol:MainnetDeploy1inch --sig "deploy1inch(uint256,uint256)" 0 1 --broadcast --rpc-url $BSC_RPC_URL --slow --account defaultKey --sender 0x48aB8AdF869Ba9902Ad483FB1Ca2eFDAb6eabe92
+FOUNDRY_PROFILE=production forge script script/forge-scripts/misc/Mainnet.Deploy.OneInchValidator.s.sol:MainnetDeployOneInchValidator --sig "deployOneInchValidator(uint256,uint256,uint256)" 0 1 0 --rpc-url $BSC_RPC_URL --slow --account defaultKey --sender 0x48aB8AdF869Ba9902Ad483FB1Ca2eFDAb6eabe92 --legacy
```
Comment (0xTimepunk) [inline script/utils/misc/run_script_mainnet_1inch.sh:19]: legacy flag not needed in BSC/AVA/POLY afaik - should remove in all instances

File: test/fuzz/scenarios/scenarios-deposit-singleDirectSingleVaultDeposit/4626.Swap.TokenInput.Slippage.OneInch.t.sol [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        actions.push(
+            TestAction({
+                action: Actions.Deposit,
+                multiVaults: false, //!!WARNING turn on or off multi vaults
+                user: 0,
+                testType: TestType.Pass,
+                revertError: "",
+                revertRole: "",
+                slippage: 1000, // 0% <- if we are testing a pass this must be below each maxSlippage,
+                dstSwap: false,
```
Comment (0xTimepunk) [inline test/fuzz/scenarios/scenarios-deposit-singleDirectSingleVaultDeposit/4626.Swap.TokenInput.Slippage.OneInch.t.sol:40]: As discussed I think we're missing:
- a Scenario with dstSwap (couldn't find it)
- a unit test showcasing rebalanceTo logic from paymaster to some address

**Review by 0xTimepunk — COMMENTED** [review]
Almost done, just minor stuff

File: src/crosschain-liquidity/1inch/OneInchValidator.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (gpersoon) [reply src/crosschain-liquidity/1inch/OneInchValidator.sol:144]: Trying to understand if further. I think `curveSwapCallback` is call to the router itself and is implemented on line 5601.
Its send tokens from the router to the caller.
Here are a few example transactions.
https://basescan.org/advanced-filter?txntype=2&age=90D&mtd=0xe413f48d%7ECurve+Swap+Callback

So this doesn't seem to be dangerous because they are all your own tokens.

File: src/crosschain-liquidity/1inch/OneInchValidator.sol (hunk 12 above)
Comment (gpersoon) [reply src/crosschain-liquidity/1inch/OneInchValidator.sol]: checking the output tokens should indeed minimize the risk

File: src/crosschain-liquidity/1inch/OneInchValidator.sol [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+                    toToken = token0;
+                } else {
+                    revert INVALID_TOKEN_PAIR();
+                }
+            }
+
+            /// @dev remap of WETH to Native if unwrapWeth flag is true
+            if (dex.shouldUnwrapWeth()) {
+                toToken = NATIVE;
+            }
```
Comment (gpersoon) [inline src/crosschain-liquidity/1inch/OneInchValidator.sol:163]: is this the only place where something is done with ETH/WETH?
currently nothing is done for Curve.
for uniswap v2, if native tokens (ETH) is used as input I think it should be converted to WETH before swapping
(it is difficult to oversee all combinations (input ETH/WETH)*(output ETH/WETH)*(3 types of swappers)*(different flags), are there enough tests to check all combinations?)
